### PR TITLE
[SPARK-9057] [STREAMING] Twitter example joining to static RDD of word sentiment values

### PR DESCRIPTION
**Pull request by Agent007**
Example of joining a static RDD of word sentiments to a streaming RDD of Tweets in order to demo the usage of the transform() method.

### CODE REVIEW

**Comment by rxin** [comment]
Thanks - can you give this pull request a more descriptive title?

**Comment by Agent007** [comment]
Done. Thanks for the feedback @rxin Let me know if there's anything else that you need.

**Comment by rxin** [comment]
cc @tdas 

Would this be more prominent & useful as an example, or more useful as a blog post?

**Comment by brennonyork** [comment]
It looks like [SPARK-9057](https://issues.apache.org/jira/browse/SPARK-9057) requires an example of `DStream.transform` done in all three primary Spark langs (java, scala, and python). This is a great start, but will need a Python and Scala example as well to be considered complete.

**Comment by brennonyork** [comment]
Is there really a need to commit the `twitter_sentiment_list.txt` file? I would think we could grab it remotely on execution rather than commit it into the source code base.

**Comment by Agent007** [comment]
Thanks for the feedback @brennonyork The Scala example was done in a previous commit on this branch. When I tried to do the same thing in Python, I got the error message: "It appears that you are attempting to broadcast an RDD or reference an RDD from an action or transformation. RDD transformations and actions can only be invoked by the driver, not inside of other transformations; for example, rdd1.map(lambda x: rdd2.values.count() * x) is invalid because the values transformation and count action cannot be performed inside of the rdd1.map transformation. For more information, see SPARK-5063."

**Comment by Agent007** [comment]
As @brennonyork suggested, the example programs now download the word-sentiment list file and `twitter_sentiment_list.txt` has been removed from the repo. The examples in all 3 languages are all in this branch. Is there anything else you gentlement need, @rxin & @tdas ?

**Comment by Agent007** [comment]
Now that 1.5 has been released, would this be ok to merge? @rxin @tdas @brennonyork Congrats on the new release, by the way!

**Comment by Agent007** [comment]
@tdas & @rxin - It was good meeting you gentlemen in person at the Meetup in Uber HQ. Thanks again for taking time to talk to me and review my code.

**Comment by rxin** [comment]
cc @tdas

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Displays the most positive hash tags by joining the streaming Twitter data with a static RDD of
+ * the AFINN word list (http://neuro.imm.dtu.dk/wiki/AFINN)
+ */
+public class JavaTwitterTagSentiments {
+    private JavaTwitterTagSentiments() {
```
Comment (tdas) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java]: Is this constructor necessary?

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Displays the most positive hash tags by joining the streaming Twitter data with a static RDD of
+ * the AFINN word list (http://neuro.imm.dtu.dk/wiki/AFINN)
+ */
+public class JavaTwitterTagSentiments {
```
Comment (tdas) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java]: This whole file should be indented with 2 spaces, to keep it consistent with other Java examples.

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        String consumerSecret = args[1];
+        String accessToken = args[2];
+        String accessTokenSecret = args[3];
+        String[] filters = Arrays.copyOfRange(args, 4, args.length);
+
+        System.setProperty("twitter4j.oauth.consumerKey", consumerKey);
+        System.setProperty("twitter4j.oauth.consumerSecret", consumerSecret);
+        System.setProperty("twitter4j.oauth.accessToken", accessToken);
+        System.setProperty("twitter4j.oauth.accessTokenSecret", accessTokenSecret);
+
```
Comment (tdas) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java]: Please add more comments throughout the code to explain stuff.

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java (hunk 1 above)
Comment (Agent007) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java]: I followed the example in https://github.com/apache/spark/blob/master/examples/src/main/java/org/apache/spark/examples/streaming/JavaKafkaWordCount.java but I can see how that may be distracting and superfluous for the purposes of showing the example. I now notice that the other Java streaming examples don't have it either. Give me a few minutes to make a new commit to remove it. Thanks for your feedback @tdas

**Comment by tdas** [comment]
Jenkins, ok to test.

**Comment by tdas** [comment]
Some high level comments. 
- I want to make it obvious that this is an example for joins on streams. So it would be good to have "join" in the name of the example files.
- We usually do not refer to external files (the AFN...) as we dont want our examples to silently break based on externat. How big is the file? If its few KBs then we can have it in our source code.

**Comment by Agent007** [comment]
@tdas The AFINN word list is 27.4 KB (see https://github.com/fnielsen/afinn/blob/master/afinn/data/AFINN-111.txt) Should I add the text file? I'll incorporate the rest of your feedback in the next commit.

**Comment by tdas** [comment]
Aah, so consider adding this to the spark/data/ directory. You will find
other data necessary for mllib tests in that.

**Comment by Agent007** [comment]
@tdas The file is now in data/streaming/ and the rest of your feedback has been incorporated. Let me know if there's anything else you need. Thanks again!

File: examples/src/main/scala/org/apache/spark/examples/streaming/TwitterHashTagJoinSentiments.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    val happiest10 = hashTags.map(hashTag => (hashTag.tail, 1))
+      .reduceByKeyAndWindow(_ + _, Seconds(10))
+      .transform{topicCount => wordSentiments.join(topicCount)}
+      .map{case (topic, tuple) => (topic, tuple._1 * tuple._2)}
+      .map{case (topic, happinessValue) => (happinessValue, topic)}
+      .transform(_.sortByKey(false))
+
+    // Print hash tags with the most positive sentiment values
+    happiest60.foreachRDD(rdd => {
+      val topList = rdd.take(10)
```
Comment (Benedicto) [inline examples/src/main/scala/org/apache/spark/examples/streaming/TwitterHashTagJoinSentiments.scala]: 10 on this line should be 60

**Comment by srowen** [comment]
@tdas this is an old one but looks good to go and finishes out your list of starter tasks for streaming.

**Comment by srowen** [comment]
@Agent007 does this require adding this large data file to the repo?

**Comment by Agent007** [comment]
Hi @srowen Yes, it has been discussed in this pull request. The AFINN word sentiment file is only 27.4 KB and is used to calculate the Twitter hash tag sentiments for the streaming example. At first, the file was downloaded at runtime as suggested by @brennonyork but @tdas wanted the file in the spark/data directory instead.

**Comment by Agent007** [comment]
@srowen Any idea why this didn't merge cleanly? I didn't see anything wrong from the Jenkins output. And, I just tried a local merge to my local copy of master and saw no merge conflicts. I've also synchronized my forked repo with the original repo, as I hadn't done that in a while. Thanks for your help.

**Comment by BenFradet** [comment]
Hi @Agent007, have you tried rebasing against master?

**Comment by srowen** [comment]
There is no merge conflict, so no rebase should be necessary. Let me try it again.

**Comment by 3ourroom** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-9057] [STREAMING] Twitter example joining to static RDD of word sentiment values (#8431)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by srowen** [comment]
@tdas I'd like to merge since the example looks good, you proposed it and looks like you substantially OK'ed it above, tests pass.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**